### PR TITLE
Use `enso.dev.insight` properly to turn Insight on

### PR DESCRIPTION
**Pull request by JaroslavTulach**
### Pull Request Description

This PR provides _development only_ (e.g. no long term) _API_ to enable (simple) usage of [GraalVM Insight](https://www.graalvm.org/latest/tools/graalvm-insight/) in the IDE and also in CLI. The insight script is sent to the internal [EPB language](https://github.com/enso-org/enso/blob/develop/engine/runtime-language-epb/src/main/java/org/enso/interpreter/epb/EpbLanguage.java#L17) responsible for interop with JavaScript, Python and other languages. Such integration revealed an inconsistency as `insight` object in the Insight script wasn't visible under the right name. Fixed by f50329a - which also changes behavior of `foreign js` - `self` is available (next to already available `this`) in instance foreign methods.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] [Insight & Enso Manual](https://github.com/enso-org/enso/pull/11385#issuecomment-2431133817) has been provided as part of this PR
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
- [x] Unit tests have been written where possible.

### CODE REVIEW

**Comment by JaroslavTulach** [comment]
# Insight & Enso Manual

Next version of the [same/similar tutorial](https://github.com/enso-org/enso/pull/14739#issuecomment-3927313595) is also available.

## Create Insight Script

Simple script to start with is for example `methods.js`:
```js
print("Initializing Insight: " + insight);
insight.on("enter", function(ctx) {
  print("Calling " + ctx.name);
}, {
  roots: true
});
```
`insight` variable is only available because of f50329a fix.

Learn more about writing **Insight Scripts**:
- [GraalVM Insight Manual](https://www.graalvm.org/latest/tools/graalvm-insight/manual/)
- GraalVM Insight [API docs](https://www.graalvm.org/tools/javadoc/org/graalvm/tools/insight/Insight.html#VERSION)

## Usage in CLI

Execute either a script or use REPL:
```bash
sbt:enso> runEngineDistribution --vm.D=enso.dev.insight=methods.js --repl
[info] Executing enso/built-distribution/enso-engine-*/enso-*/bin/enso --vm.D=enso.dev.insight=methods.js --repl
Initializing Insight: [object Object]
Calling Internal_Repl_Module___::Internal_Repl_Module___::internal_repl_entry_point___
> "Ahoj".reverse
Calling Standard.Base.Data.Text.Extensions::Standard.Base.Data.Text.Text::reverse
Calling Text.reverse.iterate
Calling Text.reverse.iterate
Calling Text.reverse.iterate
Calling Text.reverse.iterate
Calling Text.reverse.iterate
>>> johA
```

## Usage in IDE

Create an Insight script like the `methods.js` above and launch `project-manager` with additional options:
```bash
enso$ ENSO_JVM_OPTS=-Denso.dev.insight=`pwd`/methods.js ./project-manager
```
use **absolute path** to reference the Insight script file. In another terminal launch the IDE:
```bash
enso$ corepack pnpm run dev:gui
```
open a project. _Output of your Insight script_ will be displayed in the `project-manager` console whenever IDE requests the engine to execute something. Alternatively one can use `sbt` to execute the project manager:
```
enso$ ENSO_JVM_OPTS=-Denso.dev.insight=`pwd`/methods.js sbt --java-home /graalvm/ runProjectManagerDistribution
```

Vector.length

with the above project one gets following output:
```
Initializing Insight: [object Object]
Calling local.NewProject4.Main::local.NewProject4.Main::main
Calling Standard.Base.Data.Vector::Standard.Base.Data.Vector.Vector::length
Calling Standard.Base.Data.Text.Encoding::Standard.Base.Data.Text.Encoding.Encoding.type::utf_32_le
Calling Standard.Base.Data.Text.Encoding::Standard.Base.Data.Text.Encoding.Encoding::Value
Calling <inline_source>.Standard.Visualization.Helpers
Calling Standard.Visualization.Helpers::Standard.Base.Any.Any::default_visualization
Calling Standard.Visualization.Id::Standard.Visualization.Id.Id.type::table
Calling Standard.Visualization.Id::Standard.Visualization.Id.Id::Builtin
Calling Standard.Visualization.Id::Standard.Visualization.Id.Id::to_js_object
...
```
modification of the code causes re-execution of the code and produces more messages. E.g. **GraalVM Insight** is active.

### Developer's Workflow

When you make a change to the Insight script, there is **no need to restart everything**, just:
- save changes to the same Insight script
- close the IDE tab with opened project
- open the project again

The new script will be used. Enough for now or [do you want more](https://github.com/enso-org/enso/pull/11385/files#r1812209962)?

**Comment by JaroslavTulach** [comment]
### Quirks

The biggest problem right now is _multi-threaded_ execution. CLI is fine (as it is single threaded), but IDE tends to **sometimes** execute the program in parallel and [Graal.js](https://github.com/oracle/graaljs/blob/master/README.md) doesn't like it. **Graal.js** insists on enforcing _single threaded ness_. That causes issue that we need to work around:
- https://github.com/oracle/graal/pull/8266 might help - but it is not in an official release yet (as of Oct 2024)
- write your insight script in Python as [demonstrate here](https://github.com/enso-org/enso/issues/11159#issuecomment-2379235662)
- fix the _language server_ to avoid _multi threaded ness_ - we shouldn't be using _multiple thread_ anyway, our code isn't really ready for that (yet)

None of the above shall stop us from integrating this `-Denso.dev.insight` support. Quite opposite: by having it in we can start experimenting with _insights written in JavaScript_ and discover quirks that prevent us to put it into production.

## Less Damage than Expected

I was experiencing _multi threaded execution problems_ with the [simple Insight script](https://github.com/enso-org/enso/pull/11385#issuecomment-2431133817) that instruments invocations of all methods in any file, including methods in visualizations. That seems to cause issues.

However the primary use-case we have is to instrument statements in `main` function of the `Main.enso` file. E.g. something done by [Detect Assign Statements script](https://github.com/enso-org/enso/pull/11385#issuecomment-2431344514) - I haven't seen any problems while doing that. E.g. restricting the application of Insights to the user file seems to work in the IDE even right now.

File: engine/common/src/main/java/org/enso/common/ContextInsightSetup.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    }
+    var insightSrc =
+        Source.newBuilder("epb", insightFile.toFile())
+            .content(language + ":0#" + insightCode)
+            .build();
+
+    var instrument = ctx.getEngine().getInstruments().get("insight");
+    @SuppressWarnings("unchecked")
+    var insight = (Function<Source, AutoCloseable>) instrument.lookup(Function.class);
+    insightHandle = insight.apply(insightSrc);
```
Comment (JaroslavTulach) [inline engine/common/src/main/java/org/enso/common/ContextInsightSetup.java:100]: We could detect changes to the `insightFile` via [filesystem watcher](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/nio/file/WatchService.html) and re-initialize the Insight script automatically. Just:
- `insightHandle.close()` to disable the currently _active insights_
- load the new content
- `insight.apply(theNewSrc)`

that way the speed of development would be closer to JavaScript development experience.

**Comment by JaroslavTulach** [comment]
## Useful Scripts: Detect Assign Statements

Following script attaches to `main` function statements and prints out values of all local variables that are initialized together with location in the Enso code and information about just assigned local variable:
```js
print("Initializing Insight: " + insight);

insight.on("return", function(ctx, frame) {
  let equals = ctx.characters.indexOf('\n') == -1 ? ctx.characters.indexOf('=') : -1;
  let text = equals == -1 ? "" : `variable ${ctx.characters.substring(0, equals)} has just assigned`;
  print(`At ${ctx.name}:${ctx.line} ${text}`);
  for (let p in frame) {
    let v = frame[p]
    if (v != null) {
      print(` property ${p} has value ${v}`);
    }
  }
}, {
  statements: true,
  rootNameFilter: ".*main"
});
```
a sample output may look like:
```
At local.NewProject4.Main::local.NewProject4.Main::main:12 variable vector1  has just assigned
 property vector1 has value 1,2,32,13
At local.NewProject4.Main::local.NewProject4.Main::main:13 variable integer1  has just assigned
 property integer1 has value 4
 property vector1 has value 1,2,32,13
At local.NewProject4.Main::local.NewProject4.Main::main:14 variable node1  has just assigned
 property node1 has value 11,12,42,23
 property integer1 has value 4
 property vector1 has value 1,2,32,13
At local.NewProject4.Main::local.NewProject4.Main::main:15 variable node2  has just assigned
 property node1 has value 11,12,42,23
 property integer1 has value 4
 property node2 has value 22,24,84,46
 property vector1 has value 1,2,32,13
```

File: engine/common/src/main/java/org/enso/common/ContextInsightSetup.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+  private void initialize() throws IOException {
+    var insightCode = Files.readString(insightFile);
+    var language = "js";
+    if (insightFile.getFileName().toString().endsWith(".py")) {
+      language = "python";
+    }
+    var insightSrc =
+        Source.newBuilder("epb", insightFile.toFile())
+            .content(language + ":0#" + insightCode)
```
Comment (Akirathan) [inline engine/common/src/main/java/org/enso/common/ContextInsightSetup.java:94]: Where is this cryptic `:0#` string constant comming from? What does it mean?

**Review by jdunkerley — APPROVED** [review]
approving the libs bit

**Review by 4e6 — APPROVED** [review]

**Comment by 4e6** [comment]
Let's merge it!

File: engine/common/src/main/java/org/enso/common/ContextInsightSetup.java (hunk 2 above)
Comment (JaroslavTulach) [reply engine/common/src/main/java/org/enso/common/ContextInsightSetup.java:94]: `epb` is just an internal language. As such the documentation is provided on the level of tests https://github.com/enso-org/enso/pull/11385/files#diff-208ccc44a38434f59f03b9e2f77d891939bcd11de260a8a9c00c2e652fa3df5dR25 and sibling source code only.

Comment (JaroslavTulach) [reply engine/common/src/main/java/org/enso/common/ContextInsightSetup.java:94]: The support is primarily for `foreign` functions. Language is language. The number is the line number where the script is located in the original file. The text behind `#` is the code of the `foreign` function.

We are re-using `epb` as it has special support for invoking JavaScript in its own _inner context_ to workaround the single threaded limitation.

**Pull request merged**